File: androidApp/src/main/java/com/example/oradore/android/AppViewModel.kt
package com.example.oradore.android

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.oradore.api.GroupProgramEntries
import com.example.oradore.api.NetworkApi
import com.example.oradore.models.ProgramEntry
import com.example.oradore.models.Room
import com.example.oradore.models.Speaker
import com.example.oradore.storage.FavoritesStorage
import kotlinx.coroutines.launch

class AppViewModel() : ViewModel() {

    lateinit var api: NetworkApi
    lateinit var storage: FavoritesStorage

    var screen by mutableStateOf(0)

    var programEntries by mutableStateOf(emptyList<GroupProgramEntries>())
        private set

    private var favoriteProgramEntries by mutableStateOf(setOf<String>())

    var rooms by mutableStateOf(emptyList<Room>())
        private set

    var speakers by mutableStateOf(emptyList<Speaker>())
        private set

    fun init() {
        favoriteProgramEntries = storage.getFavorites().toSet()
    }

    fun fetchProgramEntries() {
        viewModelScope.launch {
            programEntries = api.fetchProgramEntries()
        }
    }

    fun programEntryById(id: String): ProgramEntry? {
        programEntries.forEach { group ->
            group.entries.forEach { entry ->
                if (entry.id == id)
                    return entry
            }
        }
        return null
    }

    fun roomById(id: String) =
        rooms.find { it.id == id }

    fun isFavorite(programEntryId: String) =
        favoriteProgramEntries.contains(programEntryId)

    private fun fav(programEntryId: String) {
        favoriteProgramEntries = favoriteProgramEntries + programEntryId
        storage.saveFavorites(favoriteProgramEntries.toList())
    }

    private fun unFav(programEntryId: String) {
        favoriteProgramEntries = favoriteProgramEntries - programEntryId
        storage.saveFavorites(favoriteProgramEntries.toList())
    }

    fun toggleFav(programEntryId: String) {
        if (isFavorite(programEntryId))
            unFav(programEntryId)
        else
            fav(programEntryId)
    }

    fun fetchRooms() {
        viewModelScope.launch {
            rooms = api.fetchRooms()
        }
    }

    fun fetchSpeakers() {
        viewModelScope.launch {
            speakers = api.fetchSpeakers()
        }
    }

    fun speakerById(id: String) =
        speakers.find { it.id == id }

    fun filterProgramEntries(where: (ProgramEntry) -> Boolean): List<GroupProgramEntries> {
        val filtered = mutableListOf<GroupProgramEntries>()
        programEntries.forEach { group ->
            val newEntries = group.entries.filter(where)
            if (newEntries.isNotEmpty()) {
                filtered += group.copy(entries = newEntries)
            }
        }
        return filtered
    }

    fun favorites(): List<GroupProgramEntries> =
        filterProgramEntries { isFavorite(it.id) }
}

File: iosApp/iosApp/AppViewModel.swift
//
//  AppViewModel.swift
//  iosApp
//
//  Created by Anja on 03.09.22.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import shared


class AppViewModel: ObservableObject, ViewModelObserver {

    @Published
    private(set) var favorites : [GroupProgramEntries] = []
    
    @Published
    private(set) var programEntries: [GroupProgramEntries] = []
    
    @Published
    private(set) var speakers: [Speaker] = []
    
    @Published
    private(set) var rooms: [Room] = []
    
    let viewModel: SharedViewModel
    
    init(api: NetworkApi, storage: FavoritesStorage) {
        self.viewModel = SharedViewModel(api: api, storage: storage)
        self.viewModel.registerObserver(observer: self)
    }
    
    deinit {
        self.viewModel.unregisterObserver(observer: self)
    }
    
    func updateFavorites(favorites: [GroupProgramEntries]) {
        self.favorites = favorites
    }
    
    func updateProgramEntries(entries: [GroupProgramEntries]) {
        self.programEntries = entries
    }
    
    func updateRooms(rooms: [Room]) {
        self.rooms = rooms
    }
    
    func updateSpeakers(speakers: [Speaker]) {
        self.speakers = speakers
    }
    
    func fetchProgramEntries() {
        viewModel.fetchProgramEntries()
    }
    
    func fetchSpeakers() {
        viewModel.fetchSpeakers()
    }
    
    func fetchRooms() {
        viewModel.fetchRooms()
    }
    
    func isFavorite(entryId: String) -> Bool {
        viewModel.isFavorite(programEntryId: entryId)
    }
    
    func toggleFav(programEntry: ProgramEntry) {
        viewModel.toggleFav(programEntryId: programEntry.id)
    }
    
    func getFavStateIcon(entryId: String) -> String {
        var iconName = ""
        if (isFavorite(entryId: entryId)) {
            iconName = "star.fill"
        } else {
            iconName = "star"
        }
        return iconName
    }
  
    func filterProgramEntries(`where` predicate: @escaping (ProgramEntry) -> Bool) -> [GroupProgramEntries] {
        viewModel.filterProgramEntries { predicate($0) ? true : false }
    }
}
